100% Clippy

Nemo157 opened the pull request:
I think merging all these changes is only worth it if we start gating CI on Clippy, otherwise the buggy allows should be dropped (just means dropping the last 2 commits). ~~With this and #1111 the deny warnings job could be changed to `cargo clippy -- -D warnings` and hit all builtin + Clippy warnings on all the crates at once.~~ (EDIT: Done since #1111 is merged)

I think the `needless_lifetimes` one could be fixed in Clippy relatively easily, presumably it just doesn't support `arbitrary_self_types` properly. The `redundant_closures` one seems like it won't be able to get a fix soon, it would need Clippy to somehow detect side-effect producing code.

I'm opening this mostly for discussion on whether gating CI on Clippy is of interest to the team. If it is then I might take a look at doing the upstream fixes to reduce the bad parts of this.

Nemo157 commented on futures-executor/src/local_pool.rs:
```diff
@@ -205,6 +205,12 @@ impl LocalPool {
     }
 }
 
+impl Default for LocalPool {
```
I could see arguments for not implementing `Default` for `LocalPool`, if so I think having an explicit `allow` for it is good documentation that it has been considered and decided against.

MajorBreakfast replied:
I can't think of any such arguments ^^' Could you list some?

Nemo157 replied:
Mostly that a `LocalPool` isn’t a data structure, so being able to create a default in a generic context isn’t that useful. IMO it’s still worth having the impl anyway though.

MajorBreakfast commented on futures-util/src/sink/with.rs:
```diff
@@ -58,7 +58,8 @@ enum State<Fut, T> {
-    fn as_pin_mut<'a>(
+    #[allow(needless_lifetimes, wrong_self_convention)] // https://github.com/rust-lang/rfcs/pull/2362#issuecomment-406804183
+    fn as_pin_mut(
         self: PinMut<'a, Self>
```
missing trailing comma

MajorBreakfast commented:
I'm in favor of merging this. We can always remove it again from the CI if it proves to be a hassle. However, I'm optimistic that we'll benefit from having it in there.

cramertj commented on futures-util/src/io/mod.rs:
````diff
@@ -288,7 +288,7 @@ pub trait AsyncWriteExt: AsyncWrite {
     /// # Ok::<(), Box<std::error::Error>>(()) }).unwrap();
     /// ```
-    fn write_all<'a>(&'a mut self, buf: &'a [u8]) -> WriteAll<'a, Self> {
+    fn write_all(&'a mut self, buf: &'a [u8]) -> WriteAll<'a, Self> {
````
Was it intentional to use in-band-lifetimes here?

Nemo157 replied:
Yep, as far as I’m aware they’re still on track to be stable in 2018 Edition. If you’ve got some reason not to use them I can revert these changes.

MajorBreakfast replied:
@cramertj My local waker opt PR you just merged also [had an in-band lifetime](https://github.com/rust-lang-nursery/futures-rs/blob/476fc37b05c9387d8d8604e03f3a11aea1ec0f3f/futures-util/src/stream/futures_unordered/node.rs#L101) in it 😇

cramertj commented on futures-util/src/sink/mod.rs:
```diff
@@ -212,7 +212,7 @@ pub trait SinkExt: Sink {
     /// `stream.forward(sink)`. The returned future will exhaust all items from
     /// `stream` and send them to `self`.
-    fn send_all<'a, St>(
+    fn send_all<St>(
```
Ditto WRT in-band lifetimes

cramertj commented on futures-util/src/sink/with.rs:
```diff
@@ -58,7 +58,8 @@ enum State<Fut, T> {
 impl<Fut, T> State<Fut, T> {
-    fn as_pin_mut<'a>(
+    #[allow(needless_lifetimes, wrong_self_convention)] // https://github.com/rust-lang/rfcs/pull/2362#issuecomment-406804183
+    fn as_pin_mut(
```
And again

cramertj commented on futures-util/src/sink/with_flat_map.rs:
```diff
@@ -60,7 +60,8 @@ where
     /// Get a pinned mutable reference to the inner sink.
-    pub fn get_pin_mut<'a>(self: PinMut<'a, Self>) -> PinMut<'a, Si> {
+    #[allow(needless_lifetimes)] // https://github.com/rust-lang/rfcs/pull/2362#issuecomment-406804183
+    pub fn get_pin_mut(self: PinMut<'a, Self>) -> PinMut<'a, Si> {
```
and again

cramertj commented on futures-util/src/stream/buffer_unordered.rs:
```diff
@@ -84,7 +84,8 @@ where
     /// stream which may otherwise confuse this combinator.
-    pub fn get_pin_mut<'a>(self: PinMut<'a, Self>) -> PinMut<'a, St> {
+    #[allow(needless_lifetimes)] // https://github.com/rust-lang/rfcs/pull/2362#issuecomment-406804183
+    pub fn get_pin_mut(self: PinMut<'a, Self>) -> PinMut<'a, St> {
```
etc...

MajorBreakfast commented:
@Nemo157 It seems that my waker reference optimization PR triggered a clippy warning that broke the CI

Nemo157 commented:
Oh, interesting, I didn't realise that Travis tested the merge commit. I was confused how those changes made it into this build. Rebased and fixed the new  warning.

MajorBreakfast commented:
@Nemo157 I didn't know that as well. Interesting!

cramertj commented:
@MajorBreakfast Yup, seems fine to me! I just wanted to check that it was an intentional decision.

MajorBreakfast commented:
I merge this with the intention to remove the `#[allow(needless_lifetimes)` annotations once `'_` lifetimes are supported in arbitrary self types. In the meantime these annotations cane be happily in there 🙂
